### PR TITLE
Run Rust tests and cache Cargo+yarn dependencies in CI

### PR DESCRIPTION
**Pull request by wbinnssmith**
This adds `cargo fetch --locked` (similar to yarn's `--frozen-lockfile`) and `cargo test --all` to our Azure pipelines runs.

This also caches Cargo and yarn's directories between runs to avoid redownloading when packages have not changed.

Test Plan:
* [x] Verify in pipelines that Cargo tests are run
* [x] Verify in pipelines that caching functions correctly

### CODE REVIEW

**Comment by mischnic** [comment]
As far as I know, `cargo check` does something like type checking (so build but without the output, and might be rather slow). Maybe use `cargo update --frozen`?

File: azure-pipelines-template.yml [hunk 1]
```diff
@@ -41,7 +45,28 @@ jobs:
...
       - script: node scripts/link-native.js
         displayName: 'Link native dependencies'
+      # From https://docs.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops
+      - task: Cache@2
+        key: 'cargo | "$(Agent.OS)" | Cargo.lock'
+        restoreKeys: |
+          cargo | "$(Agent.OS)"
+          cargo
+        path: $(Pipeline.Workspace)/target
+        displayName: Cache yarn directory
```
Comment (mischnic) [inline azure-pipelines-template.yml]:
```suggestion
        displayName: Cache Cargo directory
```

**Comment by parcel-benchmark** [comment]
## Benchmark Results
<details><summary>Kitchen Sink ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 3.59s | -139.00ms |
| Cached | 453.00ms | -27.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 401.00ms | -43.00ms 🚀 |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 502.00ms | -760.00ms 🚀 |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 1.15s | -115.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/legacy/parcel.d5807e82.webp | 102.94kb | +0.00b | 67.00ms | -17.00ms 🚀 |
| dist/modern/parcel.d5807e82.webp | 102.94kb | +0.00b | 89.00ms | +27.00ms ⚠️ |
| dist/legacy/index.76ce4b0c.js | 1.12kb | +0.00b | 90.00ms | +6.00ms ⚠️ |
| dist/modern/index.b43fcc94.js | 1.12kb | +0.00b | 90.00ms | +6.00ms ⚠️ |
| dist/legacy/index.html | 701.00b | +0.00b | 90.00ms | +6.00ms ⚠️ |
| dist/modern/index.html | 701.00b | +0.00b | 89.00ms | +5.00ms ⚠️ |
| dist/legacy/index.44edca46.css | 77.00b | +0.00b | 90.00ms | +6.00ms ⚠️ |
| dist/modern/index.0bc656bc.css | 77.00b | +0.00b | 90.00ms | +6.00ms ⚠️ |

</p></details>

<details><summary>React HackerNews ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 20.25s | +408.00ms |
| Cached | 719.00ms | -40.00ms 🚀 |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/PermalinkedComment.0a19c7e4.js | 4.24kb | +0.00b | 1.01s | -85.00ms 🚀 |
| dist/UserProfile.b2960402.js | 1.78kb | +0.00b | 1.01s | -85.00ms 🚀 |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 1.01s | -85.00ms 🚀 |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.js | 477.81kb | +0.00b | 94.00ms | +6.00ms ⚠️ |
| dist/PermalinkedComment.0a19c7e4.js | 4.24kb | +0.00b | 93.00ms | +6.00ms ⚠️ |
| dist/UserProfile.b2960402.js | 1.78kb | +0.00b | 93.00ms | +6.00ms ⚠️ |
| dist/NotFound.f3269ae7.js | 542.00b | +0.00b | 93.00ms | +6.00ms ⚠️ |
| dist/logo.24c8bf9e.png | 274.00b | +0.00b | 54.00ms | -3.00ms 🚀 |

</p></details>

<details><summary>AtlasKit Editor ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 2.28m | -13.34s 🚀 |
| Cached | 3.24s | -3.00ms |

#### Cold Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/index.e48a72e5.js | 2.46mb | -4.00b 🚀 | 1.12m | +3.05s |
| dist/editorView.ea13a2b7.js | 588.08kb | +0.00b | 59.72s | -19.06s 🚀 |
| dist/popup.119ff481.js | 171.12kb | +0.00b | 18.39s | -1.45s 🚀 |
| dist/EmojiPickerComponent.c4358f76.js | 139.84kb | +0.00b | 10.33s | -755.00ms 🚀 |
| dist/Toolbar.0c83bb49.js | 100.68kb | +0.00b | 16.99s | -2.85s 🚀 |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 13.22s | +1.80s ⚠️ |
| dist/card.1d931b4c.js | 52.43kb | +0.00b | 18.14s | -1.68s 🚀 |
| dist/Modal.558fb5db.js | 39.59kb | +0.00b | 4.74s | +348.00ms ⚠️ |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 9.11s | -655.00ms 🚀 |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 4.73s | +341.00ms ⚠️ |
| dist/smartMediaEditor.21f91ed8.js | 16.96kb | +0.00b | 14.19s | -1.55s 🚀 |
| dist/js.8167e2d5.js | 16.55kb | +0.00b | 4.73s | +341.00ms ⚠️ |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 9.39s | +1.45s ⚠️ |
| dist/component.677fad45.js | 6.23kb | +0.00b | 3.87s | +341.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 12.57s | +1.19s ⚠️ |
| dist/EmojiPickerComponent.4991ddd4.js | 3.68kb | +0.00b | 11.32s | +3.19s ⚠️ |
| dist/png-chunks-extract.608f438d.js | 3.56kb | +0.00b | 4.73s | +341.00ms ⚠️ |
| dist/dropzone.d1e8004d.js | 3.46kb | +0.00b | 14.29s | +1.45s ⚠️ |
| dist/16.b6ca7fd0.js | 2.50kb | +0.00b | 3.42s | +301.00ms ⚠️ |
| dist/date.a69000aa.js | 1.97kb | +0.00b | 5.18s | +322.00ms ⚠️ |
| dist/images.ece03ae5.js | 1.91kb | +0.00b | 5.22s | +354.00ms ⚠️ |
| dist/16.a9ea9b41.js | 1.86kb | +0.00b | 4.88s | +357.00ms ⚠️ |
| dist/16.89d45d15.js | 1.80kb | +0.00b | 4.21s | +317.00ms ⚠️ |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 9.37s | +1.55s ⚠️ |
| dist/workerHasher.d607e0b8.js | 1.77kb | +0.00b | 14.18s | -1.35s 🚀 |
| dist/status.6595d133.js | 1.69kb | +0.00b | 5.76s | +458.00ms ⚠️ |
| dist/code.efde0f48.js | 1.61kb | +0.00b | 5.07s | +327.00ms ⚠️ |
| dist/heading6.5a6082c8.js | 1.53kb | +0.00b | 10.01s | +4.54s ⚠️ |
| dist/heading3.1c92891c.js | 1.52kb | +0.00b | 5.98s | +533.00ms ⚠️ |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 5.45s | +678.00ms ⚠️ |
| dist/16.00f25203.js | 1.47kb | +0.00b | 4.21s | +317.00ms ⚠️ |
| dist/16.02056ae6.js | 1.47kb | +0.00b | 5.45s | +678.00ms ⚠️ |
| dist/emoji.b5624cf5.js | 1.46kb | +0.00b | 5.18s | +322.00ms ⚠️ |
| dist/16.00100306.js | 1.44kb | +0.00b | 5.07s | +331.00ms ⚠️ |
| dist/16.bdd61d2d.js | 1.41kb | +0.00b | 5.07s | +331.00ms ⚠️ |
| dist/heading5.88bc6249.js | 1.41kb | +0.00b | 5.98s | +533.00ms ⚠️ |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 4.88s | +357.00ms ⚠️ |
| dist/heading2.22de70fe.js | 1.34kb | +0.00b | 5.98s | +533.00ms ⚠️ |
| dist/16.d10db5a3.js | 1.32kb | +0.00b | 4.88s | +357.00ms ⚠️ |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 5.07s | +331.00ms ⚠️ |
| dist/mention.c44b48ed.js | 1.30kb | +0.00b | 5.59s | +560.00ms ⚠️ |
| dist/heading4.a2524c70.js | 1.29kb | +0.00b | 5.98s | +533.00ms ⚠️ |
| dist/Modal.44e980e0.js | 1.29kb | +0.00b | 4.73s | +340.00ms ⚠️ |
| dist/16.acff901d.js | 1.27kb | +0.00b | 4.88s | +357.00ms ⚠️ |
| dist/component.233c7a65.js | 1.26kb | +0.00b | 3.42s | +302.00ms ⚠️ |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 4.69s | +307.00ms ⚠️ |
| dist/16.828b958e.js | 1.26kb | +0.00b | 5.45s | +678.00ms ⚠️ |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 6.26s | +998.00ms ⚠️ |
| dist/16.ec84a250.js | 1.26kb | +0.00b | 4.21s | +317.00ms ⚠️ |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 5.18s | +322.00ms ⚠️ |
| dist/quote.b6d8bffb.js | 1.25kb | +0.00b | 5.76s | +458.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 12.57s | +1.19s ⚠️ |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 5.07s | +331.00ms ⚠️ |
| dist/16.458399e0.js | 1.23kb | +0.00b | 3.42s | +301.00ms ⚠️ |
| dist/decision.c48f9f1e.js | 1.22kb | +0.00b | 5.18s | +322.00ms ⚠️ |
| dist/panel-warning.b5f3360d.js | 1.22kb | +0.00b | 5.76s | +458.00ms ⚠️ |
| dist/list.2dfe3f18.js | 1.19kb | +0.00b | 5.58s | +544.00ms ⚠️ |
| dist/16.90358f40.js | 1.18kb | +0.00b | 4.89s | +354.00ms ⚠️ |
| dist/panel-error.7454ebc0.js | 1.12kb | +0.00b | 7.84s | +2.81s ⚠️ |
| dist/panel.fc70749c.js | 1.11kb | +0.00b | 5.76s | +458.00ms ⚠️ |
| dist/table.77df779b.js | 1.10kb | +0.00b | 5.76s | +447.00ms ⚠️ |
| dist/panel-success.d5752da0.js | 1.06kb | +0.00b | 7.84s | +2.78s ⚠️ |
| dist/panel-note.6a71a1f2.js | 1.05kb | +0.00b | 7.84s | +2.81s ⚠️ |
| dist/media-picker-analytics-error-boundary.49b32f4d.js | 1023.00b | +0.00b | 18.04s | +5.89s ⚠️ |
| dist/media-card-analytics-error-boundary.9a500f44.js | 1019.00b | +0.00b | 12.57s | +874.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.b884d580.js | 1019.00b | +0.00b | 13.59s | -6.23s 🚀 |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 9.37s | +1.55s ⚠️ |
| dist/simpleHasher.2c5206fd.js | 767.00b | +0.00b | 14.18s | -1.35s 🚀 |
| dist/index.html | 119.00b | +0.00b | 2.26s | +389.00ms ⚠️ |

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/pdfRenderer.f5a3f4a7.js | 1.11mb | +0.00b | 193.00ms | -21.00ms 🚀 |
| dist/EmojiPickerComponent.c4358f76.js | 139.84kb | +0.00b | 183.00ms | -23.00ms 🚀 |
| dist/media-viewer.ebcb5e1e.js | 73.43kb | +0.00b | 193.00ms | -21.00ms 🚀 |
| dist/esm.e3419f0a.js | 27.83kb | +0.00b | 183.00ms | -23.00ms 🚀 |
| dist/component.f25dcadb.js | 22.80kb | +0.00b | 98.00ms | +5.00ms ⚠️ |
| dist/DatePicker.faca45d0.js | 21.38kb | +0.00b | 173.00ms | -22.00ms 🚀 |
| dist/workerHasher.bc6346e1.js | 11.90kb | +0.00b | 183.00ms | -24.00ms 🚀 |
| dist/component.677fad45.js | 6.23kb | +0.00b | 163.00ms | +60.00ms ⚠️ |
| dist/card.3a4028bd.js | 5.74kb | +0.00b | 193.00ms | -10.00ms 🚀 |
| dist/media-viewer.f15e56f4.js | 4.01kb | +0.00b | 193.00ms | -21.00ms 🚀 |
| dist/workerHasher.31326e89.js | 1.77kb | +0.00b | 183.00ms | -24.00ms 🚀 |
| dist/list-number.d4568b49.js | 1.69kb | +0.00b | 161.00ms | +9.00ms ⚠️ |
| dist/16.07ccad90.js | 1.51kb | +0.00b | 120.00ms | +7.00ms ⚠️ |
| dist/emoji.b5624cf5.js | 1.46kb | +0.00b | 154.00ms | +11.00ms ⚠️ |
| dist/16.fa5f411c.js | 1.36kb | +0.00b | 131.00ms | +7.00ms ⚠️ |
| dist/16.d10db5a3.js | 1.32kb | +0.00b | 131.00ms | +7.00ms ⚠️ |
| dist/16.c9db7e6b.js | 1.32kb | +0.00b | 145.00ms | +11.00ms ⚠️ |
| dist/16.015dfb9d.js | 1.26kb | +0.00b | 109.00ms | +6.00ms ⚠️ |
| dist/16.828b958e.js | 1.26kb | +0.00b | 120.00ms | +7.00ms ⚠️ |
| dist/16.b9b1643b.js | 1.26kb | +0.00b | 120.00ms | +7.00ms ⚠️ |
| dist/divider.ed7ccd2e.js | 1.26kb | +0.00b | 154.00ms | +11.00ms ⚠️ |
| dist/media-card-analytics-error-boundary.ff0177b3.js | 1.25kb | +0.00b | 193.00ms | -21.00ms 🚀 |
| dist/action.01ec8983.js | 1.24kb | +0.00b | 145.00ms | +11.00ms ⚠️ |
| dist/16.90358f40.js | 1.18kb | +0.00b | 131.00ms | +7.00ms ⚠️ |
| dist/simpleHasher.79137cb4.js | 767.00b | +0.00b | 184.00ms | -23.00ms 🚀 |
| dist/index.html | 119.00b | +0.00b | 86.00ms | -77.00ms 🚀 |

</p></details>

<details><summary>Three.js ✅</summary><p>

#### Timings

| Description | Time | Difference |
| --- | --- | --- |
| Cold | 16.82s | +41.00ms |
| Cached | 643.00ms | -13.00ms |

#### Cold Bundles

*No bundle changes detected.*

#### Cached Bundles

| Bundle | Size | Difference | Time | Difference |
| --- | --- | --- | --- | --- |
| dist/Three.js | 580.60kb | +0.00b | 83.00ms | -6.00ms 🚀 |

</p></details>

**[Click here to view a detailed benchmark overview.](https://benchmark.parceljs.org/benchmark/297168939658510861)**

**Comment by wbinnssmith** [comment]
> As far as I know, `cargo check` does something like type checking (so build but without the output, and might be rather slow). Maybe use `cargo update --frozen`?

`cargo fetch` seems to be analogous to `yarn install`

**Review by devongovett — APPROVED** [review]

**Pull request merged**